modules/systemdstates/collect.go

```diff
@@ -0,0 +1,89 @@
+package systemdstates
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/coreos/go-systemd/v22/dbus"
+)
+
+const (
+	inactive = 0
+	active   = 1
+	failed   = 2
+)
+
+func (s *SystemdStates) collect() (map[string]int64, error) {
+
+	var err error
+	conn, err := dbus.New()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	allUnits, err := conn.ListUnits()
+	if err != nil {
+		return nil, err
+	}
+
+	mx := make(map[string]int64)
+
+	units := s.filterUnits(allUnits)
+	for _, unit := range units {
+
+		ut, err := s.unitType(unit.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		chartID := fmt.Sprintf("systemd_%s_active_state", ut)
+		chart := s.charts.Get(chartID)
+		if !chart.HasDim(unit.Name) {
+			_ = chart.AddDim(&Dim{ID: unit.Name})
+		}
+
+		state := -1
+		if unit.ActiveState == "active" {
+			state = active
+		}
+		if unit.ActiveState == "inactive" {
+			state = inactive
+		}
+		if unit.ActiveState == "failed" {
+			state = failed
+		}
+		mx[unit.Name] = int64(state)
+	}
+
+	return mx, nil
+}
+
+func (s SystemdStates) filterUnits(units []dbus.UnitStatus) []dbus.UnitStatus {
+
+	filtered := make([]dbus.UnitStatus, 0, len(units))
+	for _, unit := range units {
+
+		if s.unitsMatcher.MatchString(unit.Name) && unit.LoadState == "loaded" {
+			filtered = append(filtered, unit)
+		}
+	}
+
+	return filtered
+}
+
+func (s SystemdStates) unitType(unit string) (string, error) {
+	validTypes := []string{"service", "socket", "device", "mount", "automount", "swap", "target", "path", "timer", "scope"}
+	ut := ""
+	for _, t := range validTypes {
+		if strings.HasSuffix(unit, fmt.Sprintf(".%s", t)) {
+			ut = t
+			break
+		}
+	}
+
+	if ut == "" {
+		return "", fmt.Errorf("Could not find a type for : %v", unit)
+	}
+	return ut, nil
```

Comment:
it is better to make it a function

`func extractUnitType(name string) (string, error)`

and use `switch` in there

Reply:
it's a function, do you mean method is not needed and also using a switch instead of for?

Reply:
yeah, that is what i mean, we dont need access to systemd structure to extract a unit type